File: practices/tdd/examples/java-unit-test-demo/src/test/java/com/concordusa/unittestdemo/service/RemainingCapacityTest.java
package com.concordusa.unittestdemo.service;

import org.junit.jupiter.api.Test;
import org.mockito.Mockito;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.*;

class RemainingCapacityTest {

    private RemainingCapacity tested;
    private int partySize;

    @Test
    public void whenAdmitParty_CapacityReducedByPartySize(){
        //arrange
        MutableInteger mockMutableInteger = Mockito.mock(MutableInteger.class);
        tested = new RemainingCapacity(mockMutableInteger);
        partySize = 12;

        //act
        tested.admitParty(partySize);

        //assert
        Mockito.verify(mockMutableInteger, Mockito.times(1)).subtract(eq(12));
    }

    @Test
    public void whenAdmitPartyIfAllowedEntry_PartySizesExceedingCapacityRejected(){
        //arrange
        MutableInteger mockMutableInteger = Mockito.mock(MutableInteger.class);
        tested = new RemainingCapacity(mockMutableInteger);
        partySize = 12;
        when(mockMutableInteger.getCurrentValue()).thenReturn(0);

        boolean actual = tested.admitPartyIfAllowedEntry(12);

        assertThat(actual).isEqualTo(false);
    }

    @Test
    public void whenAdmitPartyIfAllowedEntry_RejectedPartiesDoNotAffectRemainingCapacity(){
        //arrange
        MutableInteger mockMutableInteger = Mockito.mock(MutableInteger.class);
        tested = new RemainingCapacity(mockMutableInteger);
        partySize = 12;
        when(mockMutableInteger.getCurrentValue()).thenReturn(0);

        //act
        tested.admitPartyIfAllowedEntry(12);

        //assert
        verify(mockMutableInteger, times(0)).subtract(any());
    }

    @Test
    public void whenAdmitPartyIfAllowedEntry_PartiesWithinCapacityAllowedEntry(){
        //arrange
        MutableInteger mockMutableInteger = Mockito.mock(MutableInteger.class);
        tested = new RemainingCapacity(mockMutableInteger);
        partySize = 12;
        when(mockMutableInteger.getCurrentValue()).thenReturn(15);

        boolean actual = tested.admitPartyIfAllowedEntry(12);

        assertThat(actual).isEqualTo(true);
    }

    @Test
    public void whenAdmitPartyIfAllowedEntry_AcceptedPartiesDoAffectRemainingCapacity(){
        //arrange
        MutableInteger mockMutableInteger = Mockito.mock(MutableInteger.class);
        tested = new RemainingCapacity(mockMutableInteger);
        partySize = 12;
        when(mockMutableInteger.getCurrentValue()).thenReturn(15);

        //act
        tested.admitPartyIfAllowedEntry(12);

        //assert
        verify(mockMutableInteger, times(1)).subtract(eq(12));
    }

    @Test
    public void whenEnforceCapacityRequiresSpaceItIsAdded_a(){
        //arrange
        MutableInteger mockMutableInteger = Mockito.mock(MutableInteger.class);
        tested = new RemainingCapacity(mockMutableInteger);
        when(mockMutableInteger.getCurrentValue()).thenReturn(-12);

        tested.enforceCapacity_a();

        verify(mockMutableInteger, times(1)).subtract(-12);
    }

    @Test
    public void whenEnforceCapacityRequiresSpaceItIsAdded_b(){
        //arrange
        MutableInteger mockMutableInteger = Mockito.mock(MutableInteger.class);
        tested = new RemainingCapacity(mockMutableInteger);
        when(mockMutableInteger.getCurrentValue()).thenReturn(-12);

        tested.enforceCapacity_b();

        verify(mockMutableInteger, times(1)).add(12);
    }

    @Test
    public void whenEnforceCapacityRequiresSpaceItIsAdded_either(){
        //arrange
        tested = new RemainingCapacity(new MutableInteger(-12));

        //act
        tested.enforceCapacity_a();

        //assert
        assertThat(tested.getRemainingCapacity()).isEqualTo(0);
    }
}

File: tdd/examples/node-unit-test-demo/__test__/remaining-capacity.test.js
const MutableInteger = require('../src/mutable-integer').default
const RemainingCapacity = require('../src/remaining-capacity').default

test('admit party calls MutableInteger subtract', () => {
    //arrange
    let mockSubtract = jest.fn();
    let mockMutableInteger = {subtract: mockSubtract};
    let tested = new RemainingCapacity(mockMutableInteger);
    let partySize = 12;

    //act
    tested.admitParty(partySize);

    //assert
    expect(mockSubtract.mock.calls.length).toBe(1);
});

test('admit party if allowed entry rejects parties beyond capacity', () => {
    //arrange
    let mockGetCurrentValue = jest.fn();
    let mockSubtract = jest.fn();
    let mockMutableInteger = {subtract: mockSubtract, getCurrentValue: mockGetCurrentValue};
    tested = new RemainingCapacity(mockMutableInteger);
    partySize = 12;
    mockGetCurrentValue.mockReturnValueOnce(0);

    //act
    let actual = tested.admitPartyIfAllowedEntry(12);

    //assert
    expect(actual).toBe(false);
});

test('rejecting a party does not reduce capacity', () => {
    //arrange
    let mockGetCurrentValue = jest.fn();
    let mockSubtract = jest.fn();
    let mockMutableInteger = {subtract: mockSubtract, getCurrentValue: mockGetCurrentValue};
    tested = new RemainingCapacity(mockMutableInteger);
    partySize = 12;
    mockGetCurrentValue.mockReturnValueOnce(0);

    //act
    tested.admitPartyIfAllowedEntry(12);

    //assert
    expect(mockSubtract.mock.calls.length).toBe(0);
});

test('admit party if allowed entry accepts parties within capacity', () => {
    //arrange
    let mockGetCurrentValue = jest.fn();
    let mockSubtract = jest.fn();
    let mockMutableInteger = {subtract: mockSubtract, getCurrentValue: mockGetCurrentValue};
    tested = new RemainingCapacity(mockMutableInteger);
    partySize = 12;
    mockGetCurrentValue.mockReturnValueOnce(15);

    //act
    let actual = tested.admitPartyIfAllowedEntry(12);

    //assert
    expect(actual).toBe(true);
});

test('accepting a party reduces capacity', () => {
    //arrange
    let mockGetCurrentValue = jest.fn();
    let mockSubtract = jest.fn();
    let mockMutableInteger = {subtract: mockSubtract, getCurrentValue: mockGetCurrentValue};
    tested = new RemainingCapacity(mockMutableInteger);
    partySize = 12;
    mockGetCurrentValue.mockReturnValueOnce(15);

    //act
    tested.admitPartyIfAllowedEntry(12);

    //assert
    expect(mockSubtract.mock.calls.length).toBe(1);
    expect(mockSubtract.mock.calls[0][0]).toBe(12);
});

test('enforce capacity removed excess occupants (a)', () => {
    //arrange
    let mockGetCurrentValue = jest.fn();
    let mockSubtract = jest.fn();
    let mockMutableInteger = {subtract: mockSubtract, getCurrentValue: mockGetCurrentValue};
    tested = new RemainingCapacity(mockMutableInteger);
    mockGetCurrentValue.mockReturnValueOnce(-12).mockReturnValueOnce(-12);

    //act
    tested.enforceCapacity_a();

    //assert
    expect(mockSubtract.mock.calls.length).toBe(1);
    expect(mockSubtract.mock.calls[0][0]).toBe(-12);
});

test('enforce capacity removed excess occupants (b)', () => {
    //arrange
    let mockGetCurrentValue = jest.fn();
    let mockAdd = jest.fn();
    let mockMutableInteger = {add: mockAdd, getCurrentValue: mockGetCurrentValue};
    tested = new RemainingCapacity(mockMutableInteger);
    mockGetCurrentValue.mockReturnValueOnce(-12).mockReturnValueOnce(-12);

    //act
    tested.enforceCapacity_b();

    //assert
    expect(mockAdd.mock.calls.length).toBe(1);
    expect(mockAdd.mock.calls[0][0]).toBe(12);
});

test('enforce capacity removed excess occupants (either)', () => {
    //arrange
    tested = new RemainingCapacity(new MutableInteger(-12));

    //act
    tested.enforceCapacity_a();
    let actual = tested.getRemainingCapacity();

    //assert
    expect(actual).toBe(0);
});
